diff -Nur povray31.orig/source/unix/makefile povray31/source/unix/makefile
--- povray31.orig/source/unix/makefile	Sat Jun 19 03:52:24 1999
+++ povray31/source/unix/makefile	Sat Oct  9 01:58:23 1999
@@ -51,7 +51,7 @@
 #CFLAGS      = -O6 -finline-functions -ffast-math -c -ansi -m386 -DCPU=586 -DCOMPILER_VER=\".`uname`.$(CC)\" -DPOV_LIB_DIR=\"$(POVLIBDIR)\" $(SRCINC) $(LIBPNGINC) $(ZLIBINC)
 
 # Linux compiler flags, Pentium II optimized
-CFLAGS      = -O6 -finline-functions -ffast-math -c -ansi -m386 -DCPU=686 -DCOMPILER_VER=\".`uname`.$(CC)\" -DPOV_LIB_DIR=\"$(POVLIBDIR)\" $(SRCINC) $(LIBPNGINC) $(ZLIBINC)
+CFLAGS      = $(RPM_OPT_FLAGS) -finline-functions -ffast-math -c -ansi -DCOMPILER_VER=\".`uname`.$(CC)\" -DPOV_LIB_DIR=\"$(POVLIBDIR)\" $(SRCINC) $(LIBPNGINC) $(ZLIBINC)
 
 # HPUX compiler flags
 #CFLAGS      = +O2 -finline-functions -c -Aa -D_HPUX_SOURCE -DCOMPILER_VER=\".`uname`.$(CC)\" $(SRCINC) $(LIBPNGINC) $(ZLIBINC)
@@ -127,12 +127,12 @@
 #default: svga
 
 # Define the default POV-Ray library location
-POVLIBDIR = /usr/local/lib/povray31
+POVLIBDIR = /usr/share/povray31
 
 # Define the path under which you want POV-Ray files to be installed,
 # typically /usr/local/bin, though Linux distributions may wish to 
 # alter this to /usr/bin
-POVPATH = /usr/local
+POVPATH = /usr
 
 #----------------------------------------------------------------------
 # There should be no compelling reason to customize the makefile beyond
diff -Nur povray31.orig/source/unix/povrayrc povray31/source/unix/povrayrc
--- povray31.orig/source/unix/povrayrc	Sat May  1 12:54:55 1999
+++ povray31/source/unix/povrayrc	Sat Oct  9 02:11:49 1999
@@ -21,8 +21,8 @@
 ;; Search path for #include source files or command line ini files not
 ;; found in the current directory.  New directories are added to the
 ;; search path, up to a maximum of 25.
-Library_Path=/usr/local/lib/povray31			;; (+Ldir)
-Library_Path=/usr/local/lib/povray31/include
+Library_Path=/usr/share/povray31			;; (+Ldir)
+Library_Path=/usr/share/povray31/include
 
 
 ;; Image size.
