diff -dur povray-3.50c.orig/configure.in povray-3.50c/configure.in
--- povray-3.50c.orig/configure.in	2004-01-09 14:27:12.000000000 +0100
+++ povray-3.50c/configure.in	2004-01-09 14:28:47.589558330 +0100
@@ -18,8 +18,6 @@
 
 AC_CHECK_LIB(m, sin)
 AC_CHECK_LIB(z, inflate)
-AC_CHECK_LIB(vga, vga_init)
-AC_CHECK_LIB(vgagl, gl_setcontextvga)
 dnl look for ios, iostream, fstream
 dnl Moved check for libX11 until after AC_PATH_EXTRA can define
 dnl the x_libraries path it needs.
Only in povray-3.50c: configure.in~
