diff -dur povray-3.50c.orig/src/frame.h povray-3.50c/src/frame.h
--- povray-3.50c.orig/src/frame.h	2002-12-31 12:51:45.000000000 +0100
+++ povray-3.50c/src/frame.h	2004-01-07 14:37:41.637836976 +0100
@@ -55,12 +55,20 @@
 class pov_istream_class;
 class pov_ostream_class;
 
+#if defined(__x86_64__) || defined(__alpha)
+typedef unsigned u_int32 ;
+#else
 typedef unsigned long u_int32 ;
+#endif
 typedef unsigned short u_int16 ;
 typedef unsigned char u_int8 ;
 typedef unsigned char byte ;
 
+#if defined(__x86_64__) || defined(__alpha)
+typedef int int32 ;
+#else
 typedef signed long int32 ;
+#endif
 typedef signed short int16 ;
 typedef signed char int8 ;
 
diff -dur povray-3.50c.orig/src/optin.cpp povray-3.50c/src/optin.cpp
--- povray-3.50c.orig/src/optin.cpp	2003-01-04 17:06:20.000000000 +0100
+++ povray-3.50c/src/optin.cpp	2004-01-07 14:38:14.675814440 +0100
@@ -436,7 +436,7 @@
 
 static bool matches(char *v1, char *v2);
 
-POV_ISTREAM *Opts_Locate_File(char *filename, unsigned long stype, char *buffer, int err_flag, POVMSObjectPtr obj);
+POV_ISTREAM *Opts_Locate_File(char *filename, u_int32 stype, char *buffer, int err_flag, POVMSObjectPtr obj);
 
 
 /*****************************************************************************
Only in povray-3.50c/src: optin.cpp~
