diff -urN povray-3.50a.org/src/optout.h povray-3.50a/src/optout.h
--- povray-3.50a.org/src/optout.h	Thu Oct 10 09:48:24 2002
+++ povray-3.50a/src/optout.h	Thu Oct 10 09:49:07 2002
@@ -46,8 +46,7 @@
 #define OFFICIAL_VERSION_NUMBER 350
 
 #define DISTRIBUTION_MESSAGE_1 "This is an unofficial version compiled by:"
-#error You must complete the following DISTRIBUTION_MESSAGE macro
-#define DISTRIBUTION_MESSAGE_2 " FILL IN NAME HERE........................."
+#define DISTRIBUTION_MESSAGE_2 " PLD/Linux Team <feedback@pld.org.pl>"
 #define DISTRIBUTION_MESSAGE_3 " The POV-Ray Team(tm) is not responsible for supporting this version."
 
 /* Number of help pages (numbered 0 to MAX_HELP_PAGE). */
