diff -dur povray-3.50c.orig/configure.in povray-3.50c/configure.in
--- povray-3.50c.orig/configure.in	2002-12-31 19:22:43.000000000 +0100
+++ povray-3.50c/configure.in	2004-01-07 14:47:04.730233920 +0100
@@ -18,9 +18,6 @@
 
 AC_CHECK_LIB(m, sin)
 AC_CHECK_LIB(z, inflate)
-AC_CHECK_LIB(png, png_get_error_ptr)
-AC_CHECK_LIB(jpeg, jpeg_std_error)
-AC_CHECK_LIB(tiff, TIFFSetWarningHandler)
 AC_CHECK_LIB(vga, vga_init)
 AC_CHECK_LIB(vgagl, gl_setcontextvga)
 dnl look for ios, iostream, fstream
@@ -30,7 +27,10 @@
 dnl Checks for header files.
 AC_PATH_XTRA
 dnl Now we can check for libX11
-AC_CHECK_LIB(X11, XFlush, LIBS="$LIBS -lX11", true, -L$x_libraries -lX11)
+AC_CHECK_LIB(X11, XFlush, LIBS="$LIBS -L$x_libraries -lX11", true, -L$x_libraries -lX11)
+AC_CHECK_LIB(png, png_get_error_ptr)
+AC_CHECK_LIB(jpeg, jpeg_std_error)
+AC_CHECK_LIB(tiff, TIFFSetWarningHandler)
 AC_HEADER_STDC
 AC_CHECK_HEADERS(limits.h sys/time.h unistd.h, )
 
Only in povray-3.50c: configure.in~
